framework/aster-frame/src/sync/wait.rs

```diff
@@ -39,68 +33,61 @@ impl WaitQueue {
     ///
     /// By taking a condition closure, his wait-wakeup mechanism becomes
     /// more efficient and robust.
-    pub fn wait_until<F, R>(&self, cond: F) -> R
+    pub fn wait_until<F, R>(&self, mut cond: F) -> R
     where
         F: FnMut() -> Option<R>,
     {
-        self.do_wait(cond, None).unwrap()
-    }
+        if let Some(res) = cond() {
+            return res;
+        }
 
-    /// Wait until some condition returns Some(_), or a given timeout is reached. If
-    /// the condition does not becomes Some(_) before the timeout is reached, the
-    /// function will return None.
-    pub fn wait_until_or_timeout<F, R>(&self, cond: F, timeout: &Duration) -> Option<R>
-    where
-        F: FnMut() -> Option<R>,
-    {
-        self.do_wait(cond, Some(timeout))
+        let (waiter, waker) = Waiter::new_pair();
+
+        loop {
+            // Enqueue the waker before checking `cond()` to avoid races
+            self.enqueue(waker.clone());
+
+            if let Some(res) = cond() {
+                return res;
+            };
+            waiter.wait();
+        }
     }
 
-    fn do_wait<F, R>(&self, mut cond: F, timeout: Option<&Duration>) -> Option<R>
+    /// Wait until some condition becomes true or cancel condition become true.
+    ///
+    /// This method works almost the same with [`WaitQueue::wait_until`], except the
+    /// waiter will be canceled if the cancel condition become true.
+    pub fn wait_until_or_cancelled<F, R, FCancel>(
```

Comment:
Well, I think accepting an waiter as a parameter may be acceptable, since it only used for wait_timeout. But since the waiter cannot be `None`, just use `Waiter` instead of `Option<Waiter>`.

And add `#[doc(hidden)]` and more comments to this method, it should not be used by other code other than wait timeout.

Reply:
I don't really object to mark this function as `#[doc(hidden)]`. But the semantics of this function make enough sense to me, so I'd like to ask if you have any concerns about the semantics. Do you think this method could be misused, or do you just think it's useless?

Reply:
I originally thought that the API might be used in scenarios other than timeouts, where there could potentially be other cancel conditions. I have not been aware of any such scenarios, but if there are, I believe the API documentation should not be hidden.

Reply:
I'm a bit afraid that this API may be unused. 

As for the currenly usage,  this API not only needs to provide a `cancel` closure, but also an additional `Waiter` parameter, which we don't want to expose to the user of `WaitQueue`.

We want that the user of `WaitQueue` is never aware of the existence of `Waiter`. `Waiters` are the internal details. 

Of course, if there are more places using this API, it is acceptable to not hide this API.

Reply:
OK. I can agree that not revealing implementation details is a good idea, so your argument makes sense to me.